packages/select/src/index.mts

```diff
@@ -25,99 +25,86 @@ type Choice<Value> = {
   type?: never;
 };
 
+type Item<Value> = Separator | Choice<Value>;
+
+const selectable = <Value,>(item: Item<Value>): item is Choice<Value> =>
+  !Separator.isSeparator(item) && !item.disabled;
+
+const render = <Value,>({ item, active }: { item: Item<Value>; active: boolean }) => {
+  if (Separator.isSeparator(item)) {
+    return ` ${item.separator}`;
+  }
+
+  const line = item.name || item.value;
+  if (item.disabled) {
+    const disabledLabel =
+      typeof item.disabled === 'string' ? item.disabled : '(disabled)';
+    return chalk.dim(`- ${line} ${disabledLabel}`);
+  }
+
+  const color = active ? chalk.cyan : (x: string) => x;
+  const prefix = active ? figures.pointer : ` `;
+  return color(`${prefix} ${line}`);
+};
+
 type SelectConfig<Value> = PromptConfig<{
   choices: ReadonlyArray<Choice<Value> | Separator>;
   pageSize?: number;
 }>;
 
-function isSelectableChoice<T>(
-  choice: undefined | Separator | Choice<T>,
-): choice is Choice<T> {
-  return choice != null && !Separator.isSeparator(choice) && !choice.disabled;
-}
-
 export default createPrompt(
   <Value extends unknown>(
     config: SelectConfig<Value>,
     done: (value: Value) => void,
   ): string => {
-    const { choices } = config;
+    const { choices: items, pageSize } = config;
     const firstRender = useRef(true);
-
     const prefix = usePrefix();
     const [status, setStatus] = useState('pending');
-    const [cursorPosition, setCursorPos] = useState(() => {
-      const startIndex = choices.findIndex(isSelectableChoice);
-      if (startIndex < 0) {
+    const [active, setActive] = useState<number>(() => {
+      const selected = items.findIndex(selectable);
+      if (selected < 0)
         throw new Error(
           '[select prompt] No selectable choices. All choices are disabled.',
         );
-      }
-
-      return startIndex;
+      return selected;
     });
 
     // Safe to assume the cursor position always point to a Choice.
-    const selectedChoice = choices[cursorPosition] as Choice<Value>;
+    const selectedChoice = items[active] as Choice<Value>;
 
     useKeypress((key) => {
       if (isEnterKey(key)) {
         setStatus('done');
         done(selectedChoice.value);
       } else if (isUpKey(key) || isDownKey(key)) {
-        let newCursorPosition = cursorPosition;
         const offset = isUpKey(key) ? -1 : 1;
-        let selectedOption;
-
-        while (!isSelectableChoice(selectedOption)) {
-          newCursorPosition =
-            (newCursorPosition + offset + choices.length) % choices.length;
-          selectedOption = choices[newCursorPosition];
-        }
-
-        setCursorPos(newCursorPosition);
+        let next = active;
+        do {
+          next = (((next + offset) % items.length) + items.length) % items.length;
```

Comment:
Why did you modify the calculation? I don't think it needs the 2 modulos. Am I missing a edge case?

Reply:
I suspect I overcorrected out of habit. Typically when we write such code, the offset could be larger than items.length, leading to a negative integer if the calculation were done differently:

Example:
Let's consider an example where items.length is 5 and next is 2:

    For an offset of -3:
        First expression: (2 - 3 + 5) % 5 = 4
        Second expression: ((2 - 3) % 5 + 5) % 5 = 4

    For an offset of -7:
        First expression: (2 - 7 + 5) % 5 = -5 % 5 = -5 (incorrect)
        Second expression: ((2 - 7) % 5 + 5) % 5 = 0
        
The second expression is how I'm used to normalizing arithmetic over an index which could result in negatives. But given the offset is in the interval [-1, 1], the simpler expression as before works. I'll revert this change.

Reply:
Also you see I was inlining the `index()` helper call in #1289 which was defined this way, so that's the side effect.